### PR TITLE
Contrib operators for object-detection bounding box related stuffs

### PR DESCRIPTION
**Pull request by zhreshold**
## Description ##
Contrib operators for object-detection bounding box related stuffs

### Essentials ###
- [x] Passed code style checking (`make lint`)
- [x] Changes are complete (i.e. I finished coding on this PR)
- [x] All changes have test coverage
- [x] For user-facing API changes, API doc string has been updated.
- [x] To my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

## Comments ##
De-coupled operators for bounding box related operations.

### CODE REVIEW

**Comment by sbodenstein** [comment]
Is the NMS op consistent with TensorFlows [`tf.image.non_max_suppression`](https://github.com/tensorflow/models/blob/master/research/object_detection/utils/np_box_list_ops.py)? And is that perhaps a better name than `box_nms`?

File: src/operator/contrib/bounding_box.cc [hunk 1]
```diff
@@ -0,0 +1,194 @@
...
+NNVM_REGISTER_OP(_contrib_box_iou)
+.describe(R"doc(Bounding box overlap of two arrays.
+  The overlap is defined as Intersection-over-Union, aka, IOU.
+  - lhs: (a_1, a_2, ..., a_n, 4) array
+  - rhs: (b_1, b_2, ..., b_n, 4) array
+  - output: (a_1, a_2, ..., a_n, b_1, b_2, ..., b_n) array
+
+  Note::
+
+    Zero gradients are back-propagated in this op for now.
```
Comment (sbodenstein) [inline src/operator/contrib/bounding_box.cc]: Are these gradients needed for any training?

**Comment by zhreshold** [comment]
@sbodenstein Functionally, yes, but not identical api. This one provide more options.
Do you have a suggestion for the name?
There's an alias "box_non_maximum_suppression"

File: src/operator/contrib/bounding_box-inl.h [hunk 2]
```diff
@@ -0,0 +1,730 @@
...
+    // apply nms
+    // go through each box as reference, suppress if overlap > threshold
+    // sorted_index with -1 is marked as suppressed
+    for (int ref = 0; ref < topk; ++ref) {
+      int num_worker = topk - ref - 1;
+      if (num_worker < 1) continue;
+      Kernel<nms_impl, xpu>::Launch(s, num_batch * num_worker, sorted_index.dptr_,
+        buffer.dptr_, areas.dptr_, num_worker, ref, num_elem, width_elem,
+        coord_start, id_index, param.overlap_thresh, param.force_suppress, param.in_format);
+    }
```
Comment (sxjscience) [inline src/operator/contrib/bounding_box-inl.h]: Does it mean that we need to launch the kernel for `topk` times?

Comment (zhreshold) [reply src/operator/contrib/bounding_box-inl.h]: Yep, that is because we need to consider broadcasting suppression problem.
If scores A > B > C, A suppress B with IOU 0.8, B suppress C with IOU 0.6, A don't suppress C with IOU 0.4. Then we should suppress B and keep C because B is already suppress by A and C has nothing to do with A.
Therefore, until we get the status of B, we can't determine the status of C.

Comment (sxjscience) [reply src/operator/contrib/bounding_box-inl.h]: OK.  There may be a possible speed problem if `topk` is large. However, the current implementation should be good in order to run the experiments. Could we later revise the kernel using similar logic as https://github.com/rbgirshick/py-faster-rcnn/blob/master/lib/nms/nms_kernel.cu?

File: src/operator/contrib/bounding_box.cc (hunk 1 above)
Comment (zhreshold) [reply src/operator/contrib/bounding_box.cc]: Gradients are rarely needed from these operations, it's a bit tricky to figure out what should be done to gradients for these kind of operators. 
TF put these into tf.image, I don't think they intended to do any form of backpropagation.

File: src/operator/contrib/bounding_box-inl.h (hunk 2 above)
Comment (zhreshold) [reply src/operator/contrib/bounding_box-inl.h]: Marked. I can later tackle on this part, for now I don't encounter perf issue with normal settings.

**Comment by zhreshold** [comment]
@piiswrong Can we merge? These ops will be used in tutorials.

**Comment by sbodenstein** [comment]
> @sbodenstein Functionally, yes, but not identical api. This one provide more options.
> Do you have a suggestion for the name? There's an alias "box_non_maximum_suppression"

If it has that alias, then that is fine.

File: src/operator/contrib/bounding_box.cc [hunk 3]
```diff
@@ -0,0 +1,182 @@
...
+.describe(R"doc(Bounding box overlap of two arrays.
+  The overlap is defined as Intersection-over-Union, aka, IOU.
+  - lhs: (a_1, a_2, ..., a_n, 4) array
+  - rhs: (b_1, b_2, ..., b_n, 4) array
+  - output: (a_1, a_2, ..., a_n, b_1, b_2, ..., b_n) array
+
+  Example::
+
+    x = [[0.5, 0.5, 1.0, 1.0], [0.0, 0.0, 0.5, 0.5]]
+    y = [0.25, 0.25, 0.75, 0.75]
```
Comment (Jing-Luo) [inline src/operator/contrib/bounding_box.cc]: Should the example be y = [[0.25, 0.25, 0.75, 0.75]]?
I tried this branch and it seemed that y should be 2d array otherwise there would be errors:
src/operator/contrib/./bounding_box-inl.h:489: Check failed: rshape.ndim() >= 2 (1 vs. 2) rhs must have dim >= 2 1 provided

**Comment by Jing-Luo** [comment]
@piiswrong @cjolivier01 why this pr hasn't been merged yet?

File: src/operator/contrib/bounding_box.cc (hunk 1 above)
Comment (piiswrong) [reply src/operator/contrib/bounding_box.cc]: Why define gradient at all?
